pdns/recursor_cache.cc

```diff
@@ -335,7 +377,8 @@ time_t MemRecursorCache::get(time_t now, const DNSName& qname, const QType qt, b
       for (auto i = entries.first; i != entries.second; ++i) {
         firstIndexIterator = map->d_map.project<OrderedTag>(i);
 
-        if (i->d_ttd <= now) {
+        // When serving stale, we consider expired records
+        if (i->d_ttd <= now && !serveStale && i->d_servedStale == 0) {
```

Comment:
It might be good to create a `isEntryUsable(const CacheEntry& entry, time_t now, bool serveStale)` function, or something like that